### DISCUSSION TITLE
AI Deck in Client mode (Connect to a Wifi Hotspot)

### DISCUSSION THREAD

**Discussion by migranram** [bitcraze/discussions#151] [Q&A] [answered]
Hello everyone!
I am trying to connect the AI Deck to a wifi network in order to be able to receive images from several drones at the same time. But unluckily the configuration of the KBuilds menuconfig is not working ("Connect to a Wifi network" instead of "Act as Access Point").
Does anyone know why it might be happening? If anybody has an idea on how to receive images from several drones in any other way, I am all ears! 
I saw a post from some time ago saying that this functionality wasn't implemented yet. But since it appears in the configuration, I do not know if that may have changed. In case not, any ideas on how to program and load the code to the ESP (which I guess is the one controlling the communication), I don't have much experience with such embedded devices.

Thank you in advance!

**Comment by knmcguire** [comment]
Hi!

I assume you have followed [this tutorial ](https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/) right? Where you flashed the .bin file of the wifi example? The thing is that that is a special binary that contains the wifi communication but works independently of the firmware state of the crazyflie itself (the define [SETUP_WIFI_AP ](https://github.com/bitcraze/aideck-gap8-examples/blob/204aaff2c183e26bf109f7c6201dd2f045916f46/examples/other/wifi-img-streamer/wifi-img-streamer.c#L178)deals with that). This is the production binary we flash on the aidecks before we ship it. 

The problem with this binary that is will not react on the state of the crazyflie firmware to setup the wifi differently, so for that you will need to  build and flash this yourself with these instructions: https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/test-functions/wifi-streamer/

Let me know if I guessed correctly or not! I do agree that the getting started tutorial is not clear about this so I'll add a warning there.

**Reply by migranram** [reply]
Hi!
Thank you for your answer. I was actually trying with the second tutorial you sent me, where I compile the wifi-streamer example too.
So firstly I configure the crazyflie with the make menuconfig to connect to a wifi network I have created, save, compile it and flash it into the drone. Then compile the wifi example and flash it into the GAP8. But the drone keeps creating a wifi network instead of connecting to one, I don't know why.
The only problem I faced with it, is that I had to flash the GAP8 first and then the STM32, because otherwise, the flashing of the GAP8 gets stuck at 0%. Maybe that is the problem. (I also tried to put the address like "radio://0/94/250" instead of "radio://0/94/250K/E7E7E7E70C", but it did not work).

Thank you!

**Reply by knmcguire** [reply]
Ah oke it is good to know that you actually tried the second guide to. Let me start a new thread for another question

**Comment by knmcguire** [comment]
If you configured the firmware to access another AP, you should get some feedback on the console tab of the CFclient, like that it is connected with an IP address (since you need that address to run the viewer with -n). 

Could you perhaps copy paste console tab print out here after a restart?

**Reply by migranram** [reply]
So if I configure the drone to connect to wifi I get this:
```
SYS: Build 129:79aead2deafe (2022.05 +129) CLEAN
SYS: I am 0x31313931353551140040001F and I have 1024KB of flash!
CFGBLK: v1, verification [OK]
DECK_CORE: 1 deck(s) found
DECK_CORE: Calling INIT on driver bcAI for deck 0
CPX-EXT-ROUTER: Waiting for CPX External router initialization
CPX-EXT-ROUTER: CPX External router initialized
AIDECK: AI-deck will connect to WiFi
MPU9250 I2C connection [OK].
AK8963 I2C connection [OK].
LPS25H I2C connection [OK].
ESTIMATOR: Using Complementary (1) estimator
CONTROLLER: Using PID (1) controller
MTR-DRV: Using brushed motor driver
SYS: About to run tests in system.c.
SYS: NRF51 version: 2022.05
EEPROM: I2C connection [OK].
STORAGE: Storage check [OK].
AK8963: Self test [OK].
DECK_CORE: Deck 0 test [OK].
SYS: Self test passed!
STAB: Wait for sensor calibration...
SYS: Free heap: 14600 bytes
STAB: Ready to fly.
STAB: WARNING: stabilizer loop rate is off (996)
```
And if configured to be accesspoint I get this:
```
SYS: Crazyflie 2.0 is up and running!
SYS: Build 129:79aead2deafe (2022.05 +129) CLEAN
SYS: I am 0x31313931353551140040001F and I have 1024KB of flash!
CFGBLK: v1, verification [OK]
DECK_CORE: 1 deck(s) found
DECK_CORE: Calling INIT on driver bcAI for deck 0
CPX-EXT-ROUTER: Waiting for CPX External router initialization
CPX-EXT-ROUTER: CPX External router initialized
AIDECK: AI-deck will become access point
MPU9250 I2C connection [OK].
AK8963 I2C connection [OK].
LPS25H I2C connection [OK].
ESTIMATOR: Using Complementary (1) estimator
CONTROLLER: Using PID (1) controller
MTR-DRV: Using brushed motor driver
SYS: About to run tests in system.c.
SYS: NRF51 version: 2022.05
EEPROM: I2C connection [OK].
STORAGE: Storage check [OK].
AK8963: Self test [OK].
DECK_CORE: Deck 0 test [OK].
SYS: Self test passed!
STAB: Wait for sensor calibration...
SYS: Free heap: 14600 bytes
STAB: Ready to fly.
STAB: WARNING: stabilizer loop rate is off (995)
```

Which seems to be responding to the configuration.

**Reply by migranram** [reply]
I actually should get a message from the GAP8 like this in the console:
```
GAP8: -- WiFi image streamer example --
GAP8: Starting camera task...
```
But it seems to get stuck at the start-up in both cases, since I don't receive anything from the GAP8. Can be seen in the start_example() function of the [example.](https://github.com/bitcraze/aideck-gap8-examples/blob/204aaff2c183e26bf109f7c6201dd2f045916f46/examples/other/wifi-img-streamer/wifi-img-streamer.c#L371)

**Reply by knmcguire** [reply]
I can indeed confirm this with the latest master of crazyflie-firmware.

I'm worried since I can not see any message of the ESP initializing like:

```
ESP32: I (910) SYS: Initialized
ESP32: I (910) DISCOVERY: Hostname is aideck-B0A2D8
```

While this works with the latest release.

**Comment by knmcguire** [comment]
I suggest that some latest commits due to PR https://github.com/bitcraze/crazyflie-firmware/pull/1081 has broken this somehow... 

Could you revert to commit b45f54fad1ee968e21b0f992bf626539a589edf5 on the crazyflie-firmware and see if it works then?

**Reply by migranram** [reply]
This seems to work fine!

**Comment by knmcguire** [comment, marked as answer]
Another, perhaps better solution, is to update the ESP as well with

```
git clone https://github.com/bitcraze/aideck-esp-firmware

docker run --rm -it -v $PWD:/module/ --device /dev/ttyUSB0 --privileged -P bitcraze/builder /bin/bash -c "openocd -f interface/ftdi/olimex-arm-usb-tiny-h.cfg -f board/esp-wroom-32.cfg -c 'adapter_khz 20000' -c 'program_esp build/bootloader/bootloader.bin 0x1000 verify' -c 'program_esp build/aideck_esp.bin 0x10000 verify reset exit'"

cfloader flash build/aideck_esp.bin deck-bcAI:esp-fw -w radio://0/80/2M
```

It seems that is necessary to match the latest changes on the firmware.

I'm currently in the process to try that out so fingers crossed!

**Reply by knmcguire** [reply]
Yes! now I get: 

```
SYS: Free heap: 14600 bytes
CPX: ESP32: I (832) SYS: Initialized
CPX: ESP32: I (832) DISCOVERY: Hostname is aideck-B0A2D8
CPX: GAP8: -- WiFi image streamer example --
CPX: ESP32: I (2752) WIFI: wifi_init_sta finished.
STAB: Ready to fly.
CPX: GAP8: Starting camera task...
CPX: ESP32: I (6282) WIFI: got ip: 192.168.6.59
CPX: ESP32: I (6282) WIFI: country: 
CPX: ESP32: I (6282) WIFI: rssi: -50
CPX: ESP32: I (6282) WIFI: 11b: 1, 11g: 1, 11n: 1, lr: 0
CPX: WiFi connected to ip: 192.168.6.59
CPX: GAP8: Wifi connected (192.168.6.59)
```

**Reply by knmcguire** [reply]
It is not ideal that the stm and esp are not backwards compatible at this moment.. but we should do another release in august which should make it easier.

now we just have to make sure to not have the aideck on the crazyflie if we flash the latest release on the cfclient (the one that updates the decks)

**Reply by migranram** [reply]
Okay, so once updated the ESP with the latest aideck-esp-firmware always take off the aideck if I need to flash the crazyflie with the cfclient again, right?
By the way: Where can I get that "bitcraze/builder" docker image to try this out?

**Reply by knmcguire** [reply]
yes exactly.. unfortunately the cfclient does not have an option to opt out yet so. 

( You could take advantage of [this bug](https://github.com/bitcraze/crazyflie-clients-python/issues/616) and attach a flowdeck after updating the esp manually)

If you have docker installed on your system, it will automatically pull that docker container from the internet before using it to build the firmware. But the readme if the aideck-esp-firmware also has other way to build and flash the esp: https://github.com/bitcraze/aideck-esp-firmware

**Reply by migranram** [reply]
Okay thank you for your help! :D

**Reply by knmcguire** [reply]
Could you perhaps mark this as answer instead, eventhough this is not what you tried in the end? 

We prefer people to update the firmwares for development instead of reverting to an earlier commit. The latest commits does include improvements for the CPX communication.

**Reply by knmcguire** [reply]
thanks!

**Comment by Kaijun101** [comment]
Hi, I am currently having issues with setting up the AI-Deck to connect to WiFi, i.e., after flashing the cf firmware the CF does not connect to the wifi or rather stays as an access point. The CF is currently using the 2022.12 firmware. I have followed the steps in : https://github.com/bitcraze/aideck-gap8-examples/blob/master/docs/test-functions/wifi-streamer.md
```
$ git clone https://github.com/bitcraze/crazyflie-firmware.git
$ cd crazyflie-firmware
$ make menuconfig
$ make
$ cfloader flash cf2.bin stm32-fw -w [CRAZYFLIE_URI]
```
Below are the output from the console:
```
SYS: ----------------------------
SYS: Crazyflie 2.1 is up and running!
SYS: Build 0:b8d2e0b9e983 (2022.12) CLEAN
SYS: I am 0x3530343431315103001C002B and I have 1024KB of flash!
CFGBLK: v1, verification [OK]
DECK_CORE: 2 deck(s) found
DECK_CORE: Calling INIT on driver bcFlow for deck 0
PMW: Motion chip id: 0x49:0xB6
DECK_CORE: Calling INIT on driver bcAI for deck 1
CPX-EXT-ROUTER: Waiting for CPX External router initialization
CPX-EXT-ROUTER: CPX External router initialized
AIDECK: Not setting up WiFi
IMU: BMI088: Using I2C interface.
IMU: BMI088 Gyro connection [OK].
IMU: BMI088 Accel connection [OK]
IMU: BMP388 I2C connection [OK]
ESTIMATOR: Using Kalman (2) estimator
CONTROLLER: Using PID (1) controller
MTR-DRV: Using brushed motor driver
SYS: About to run tests in system.c.
SYS: NRF51 version: 2022.12
EEPROM: I2C connection [OK].
STORAGE: Storage check [OK].
IMU: BMI088 gyro self-test [OK]
DECK_CORE: Deck 0 test [OK].
DECK_CORE: Deck 1 test [OK].
SYS: Self test passed!
STAB: Wait for sensor calibration...
SYS: Free heap: 11712 bytes
CPX: ESP32: I (800) SYS: Initialized
CPX: ESP32: I (800) DISCOVERY: Hostname is aideck-E0CA5C
CPX: ESP32: I (800) DISCOVERY: MAC (STA) is 9c:9c:1f:e0:ca:5c
STAB: Ready to fly.
```

**Reply by knmcguire** [reply]
Hi! Sorry, you will need to make a new discussion for this, as this discussion thread is already closed and about a different issue that you are dealing with. Also check if your issue was not handled before in the [AIdeck Q&A category](https://github.com/orgs/bitcraze/discussions/categories/q-a?discussions_q=category%3AQ%26A+label%3A%22AI+deck%22)
